### ISSUE TITLE
Ancient block sync stalls

### ISSUE THREAD

**Issue by dip239** [paritytech/parity-ethereum#7008]
> I'm running:
>
> - **Parity version**: Parity/v1.8.2-beta-1b6588c-20171025/x86_64-linux-gnu/rustc1.21.0
> - **Operating system**: Linux
> - **And installed**: sudo apt-get install ./parity_1.8.2_amd64.deb

configuration detail: chain is simlinked to another drive
```
human@EtherBox:~$ ls -lisa ./.local/share/io.parity.ethereum/chains
524382 0 lrwxrwxrwx 1 human human 28 Sep  3 21:25 ./.local/share/io.parity.ethereum/chains -> /media/human/CHAIN_DB/chains
```

After warp sync, I'm unable to get all missing blocks fetched. After new start Parity forgets the already fetched blocks and start downloading them again and again.

Here is a sample of 3 parity runs

![image](https://user-images.githubusercontent.com/4574735/32586099-1504cf62-c501-11e7-86cc-c44abe018ca3.png)

 
Here is the log file from above (as text file)
[parity-3x-sync.txt](https://github.com/paritytech/parity/files/1456141/parity-3x-sync.txt)

---

**Comment by roninkaizen** [comment]
normal behaviour/ misunderstanding-
as you looked at the screen you posted, it became obvious-
confirmed = yes
unwanted, unintended=no
normal behavior, other counting on the transfers, 
also having that, being fully sync on that nodes, 
irritiating=yes
annoyance=no

Document it, then it is known to be normal. thx

**Comment by 5chdn** [comment]
That's the ancient block download.

Warp-sync downloads the latest snapshot and the last 30k blocks. After that, it starts downloading the full blockchain (yellow numbers).

**Comment by dip239** [comment]
@5chdn @roninkaizen 
What is the reason to download ancient blocks again and again? This is the issue reported.

Please check the block numbers:
1st run: downloaded ancient blocks from #3574411 to #3614671
2st run: downloaded ancient blocks from #3573015 to #3576698    
3st run: downloaded ancient blocks from #3573269  to #3575935    

ancient blocks being downloaded in the same range again und again. 
What is the reason to do it if it is not a bug?

**Comment by arkpar** [comment]
This is really weird indeed. Never seen this before. Could you restart with `-l sync=trace` and post logs?

**Comment by dip239** [comment]
yes, please:
Here 4 subsequent runs, two last are with `-l sync=trace`.
![grafik](https://user-images.githubusercontent.com/4574735/32693194-fc569ec8-c726-11e7-83aa-a56a3d011406.png)

Trace-Logs:
[parity-trace.log.zip](https://github.com/paritytech/parity/files/1464087/parity-trace.log.zip)

**Comment by 5chdn** [comment]
Having similar issues with 1.8.2

**Comment by 5chdn** [comment]
Can we also make sure we do not purge ancient blocks whenever a warpsync kicks in?

**Comment by tomusdrw** [comment]
As a workaround:
1. Remove database
2. Warp again to latest snapshot (hopefuly warp health improves soon with new bootnodes being rolled out)
3. Download ancient blocks again.

> Can we also make sure we do not purge ancient blocks whenever a warpsync kicks in?

@5chdn That's a separate issue, can you log it?

**Comment by 5chdn** [comment]
@tomusdrw I think it is this one: #6350

**Comment by lght** [comment]
> As a workaround:
>
> Remove database
> Warp again to latest snapshot (hopefuly warp health improves soon with new bootnodes being rolled out)
> Download ancient blocks again.

:+1: for this workaround

With a 1.10.0 nightly build, warp took only ~40min from a fresh db! Syncing ancient blocks now with 25 peers (took a little time to get full peers).

Will report back with sync status, currently at block  4880220. Warp synced to block 4880000

**Comment by lght** [comment]
@arkpar is it possible that part of @dip239's issue is because of some corrupt block in the import round he was working on causing all blocks in that round to be rolled back on restart?

**Comment by 5chdn** [comment]
lght, this issue is about ancient block sync that happens _after_ the warp sync.

**Comment by GoodMirek** [comment]
The issue **Ancient block sync stalls** has happened to me two times so far. Parity and Linux version strings follow:
```
version Parity/v1.10.0-unstable-25b19835e-20180117/x86_64-linux-gnu/rustc1.22.1

Linux wedos1 4.14.13-300.fc27.x86_64 #1 SMP Thu Jan 11 04:00:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

It seems to happen once per ~1.5M blocks (1st occurrence at block 1639023, 2nd occurrence at block 2996831). At first occurrence, after the graceful restart (SIGHUP) the parity process continued to download ancient blocks from where it left off. There were parity process restarts in between the occurrences.

Second occurrence of the issue just happened and I am keeping the process at the stalled state. Blockchain head keeps syncing. If you need any further info from the stalled process or need access to the system, let me know (there is no private info on the system). If I did not hear back within 3 days, I planned to restart the process with TRACE loglevel. I expect the issue could happen again until all blocks are downloaded.

While being in stale state, I took coredump of the process using gcore. Attached is gdb backtrace [gdb_core_9187.txt](https://github.com/paritytech/parity/files/1642594/gdb_core_9187.txt). Log is at info level, so probably not useful. If you need the coredump file, I will share it (it contains no private info).

If that helps, I can rebuild parity in a different way, e.g. with symbols and without optimizations. I would appreciate a hint how to do that as I have zero rust ecosystem knowledge.

**Comment by 5chdn** [comment]
This just happened to one of my fresh 1.10.0 nodes if anyone wants to debug this.

**Comment by folsen** [comment]
@ngotchac can you double-check that #8642 addresses this and if so close the issue please.

**Comment by folsen** [comment]
@GoodMirek Thanks so much for this, it's really helpful, however I don't think it's related to this issue. I'd also say that 4GB RAM is probably not enough to run Parity with those caches, ram usage can depend a bit on peers as well although not a ton. My own node consumes about 10gb ram with similar cache sizes, though that is still a bit above expectation for various reasons. It would be a separate issue to try to investigate running parity under OpenVZ vs KVM VM, it should work, but I personally don't know the requirements on the VM side.

**Comment by GoodMirek** [comment]
@folsen The point is that on the same container with the same command line options a previous version built with rustc 1.25 ran just fine for more than a week. Though I do not remember which commit hash it was, but not more than a month old.
I know the previous build I ran suffered from this particular issue "Ancient block sync stalls", as I had to restart the process several times to complete download of old blocks.
